src/globus_sdk/experimental/globus_app/globus_app.py

```diff
@@ -0,0 +1,429 @@
+                    self._scope_requirements[resource_server], scopes
+                )
+
+        # update the app's ValidatingTokenStorage's scope requirements to enforce the
+        # updated scope requirements
+        self._validating_token_storage.scope_requirements = self._scope_requirements
+
+
+class UserApp(GlobusApp):
+    """
+    A ``GlobusApp`` for login methods that require an interactive flow with a user.
+
+    Can either use a native application client by only giving ``client_id`` or a
+    templated client by giving both ``client_id`` and ``client_secret``.
+
+    .. tab-set::
+
+        .. tab-item:: Future Example Usage
+
+            .. code-block:: python
+
+                app = UserApp("myapp", client_id=NATIVE_APP_CLIENT_ID)
+                client = TransferClient(app=app)
+                res = client.operation_ls(COLLECTION_ID)
+
+    """
+
+    _login_client: NativeAppAuthClient | ConfidentialAppAuthClient
+    _authorizer_factory: (  # type:ignore
+        AccessTokenAuthorizerFactory | RefreshTokenAuthorizerFactory
+    )
+
+    def __init__(
+        self,
+        app_name: str,
+        *,
+        scope_requirements: dict[str, list[Scope]] | None = None,
+        client_id: UUIDLike | None = None,
+        client_secret: str | None = None,
+        config: GlobusAppConfig | None = None,
+    ):
+        """
+        :param app_name: A string to identify this app. Used for default tokenstorage
+            location and in the future will be used to set user-agent when this app is
+            attached to a service client
+        :param scope_requirements: A dict of lists of required scopes indexed by
+            their resource server.
+        :param client_id: A UUID for the Globus client this app will use. This value
+            is currently required in all cases.
+        :param client_secret: If using a templated client, the value of its client
+            secret. If omitted or passed as None the app is assumed to be a native
+            app.
+        :config: A ``GlobusAppConfig`` used to control various behaviors of this app.
+        """
+
+        # in the future we might use an environment variable if this is passed as None,
+        # but for now client_id is required
+        if client_id is None:
+            raise ValueError("client_id is required")
+
+        self.client_id = client_id
+        self.client_secret = client_secret
+        super().__init__(app_name, scope_requirements=scope_requirements, config=config)
+
+        # either get config's LoginFlowManager
+        # or make a default CommandLineLoginFlowManager
+        if self.config.login_flow_manager:
+            self._login_flow_manager = self.config.login_flow_manager
+        else:
+            self._login_flow_manager = CommandLineLoginFlowManager(
+                self._login_client,
+                refresh_tokens=self.config.refresh_tokens,
+            )
```

Comment:
The config interface for login flow manager doesn't really expose the ability to override default behavior in a somewhat confusing way:

* A GlobusApp will always construct a login client (which it holds privately).
* If `config.login_flow_manager` was omitted, that login client will be passed to the default login flow manager (CommandLineLoginFlowManager)
* `config.login_flow_manager` can be overridden but cannot have access to the login client created by the app.

So, for instance, if I write my own custom `CarrierPigeonLoginFlowManager` which communicates information back and forth by carrier pigeons there's no way to have the login client associated my globusapp associated with that loginflowmanager.

Not sure if there's a simple fix to align this overrideability (I also see that refreh_token configuration comes through only for the CommandLineLoginFlowManager). Maybe we could:
* Accept a class instead of an actual loginflowmanager in the config?
* Accept a provider which takes a pre-canned set of a parameters & create a loginflowmanager?
* Make the login client association be something that happens post-instantiation for a `LoginFlowManager`?

Reply:
I'll take a pass at improving this, but might need to discuss further.

I also want to note that `CarrierPigeonLoginFlowManager` made me smile ^^

Reply:
Sure, I'm happy to jump on a huddle if it'd be easier to try to find a path through this synchronously; just ping me. 

`CarrierPigeonLoginFlowManager` is great for those times when logins as happening just too dang fast!